### PR TITLE
Web implementation

### LINKED ISSUE #86: Missing web implementation
As Flutter web is now seeing major adoption and has already been in the **beta** stage for quite some time, Flutter web support is a good idea.

Anyways, this plugin is currently missing web support and that would be great to have 😃

### PR DESCRIPTION
**Pull request by deepak786**
Following functions are not yet supported on Web:

- unreadConversationCount
- setInAppMessagesVisibility
- displayHelpCenter
- sendTokenToIntercom
- handlePushMessage
- isIntercomPush
- handlePush

because there is no corresponding Intercom javascript API.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -110,3 +111,28 @@ Future<dynamic> backgroundMessageHandler(Map<String, dynamic> message) async {
...
+
+    window.intercomSettings = {
+        app_id: APP_ID,
+        hide_default_launcher: true, // set this to false, if you want to show the default lanuncher
+    };
+    (function(){var w=window;var ic=w.Intercom;if(typeof ic==="function"){ic('reattach_activator');ic('update',w.intercomSettings);}else{var d=document;var i=function(){i.c(arguments);};i.q=[];i.c=function(args){i.q.push(args);};w.Intercom=i;var l=function(){var s=d.createElement('script');s.type='text/javascript';s.async=true;s.src='https://widget.intercom.io/widget/' + APP_ID;var x=d.getElementsByTagName('script')[0];x.parentNode.insertBefore(s, x);};if(document.readyState==='complete'){l();}else if(w.attachEvent){w.attachEvent('onload',l);}else{w.addEventListener('load',l,false);}}})();
+
+</script>
+```
+#### Following functions are yet not supported or work on Web:
```
Comment (creativecreatorormaybenot) [inline README.md]:
```suggestion
#### Following functions are not yet supported on Web:
```

**Comment by v3rm0n** [comment]
Looks interesting. Can you add some tests as well?

**Comment by deepak786** [comment]
@v3rm0n ok, I will add the tests.

@Zazo032 @v3rm0n I think the enum ```IntercomVisibility``` should be moved to ```intercom_flutter_platform_interface.dart```. What do you think?

**Comment by Zazo032** [comment]
> @v3rm0n ok, I will add the tests.
> 
> @Zazo032 @v3rm0n I think the enum `IntercomVisibility` should be moved to `intercom_flutter_platform_interface.dart`. What do you think?

[FlutterFire](https://github.com/FirebaseExtended/flutterfire/search?q=enum) and [official Flutter plugins](https://github.com/flutter/plugins/search?q=enum) have enums on the interface package, not sure if there's any official guidance on that specific topic. I guess interface is the way to go unless there are different values for each platform. Even in that case, some values could be skipped on some specific platform.

**Comment by deepak786** [comment]
> [FlutterFire](https://github.com/FirebaseExtended/flutterfire/search?q=enum) and [official Flutter plugins](https://github.com/flutter/plugins/search?q=enum) have enums on the interface package, not sure if there's any official guidance on that specific topic. I guess interface is the way to go unless there are different values for each platform. Even in that case, some values could be skipped on some specific platform.

I think we should move the enum to the platform interface but it should be mentioned in the changelog because, after this, the users will get the error for IntercomVisibility enum.

**Comment by deepak786** [comment]
@Zazo032 have you checked and tested the web implementation?

One question, As some functions are not supported on web. So if that functions will be called then the user will see the unimplemented error.
Instead of throwing the unimplemented error, we can just return a valid string that "this function is not supported on the web" by overriding those functions.

What do you think?

**Comment by IchordeDionysos** [comment]
Usually, platform interface and implementation are split into different packages 🤔

**Comment by creativecreatorormaybenot** [comment]
Agree with @IchordeDionysos :) See [this article](https://medium.com/flutter/how-to-write-a-flutter-web-plugin-part-2-afdddb69ece6) to see how to properly set up the web implementation.

However, **technically** it does not matter because every Dart file is a separate library.

**Comment by IchordeDionysos** [comment]
However, it does make it easier to discover platform_interfaces and publishing implementations for diffent platforms independently of the package author or to override the implementation for one platform.

**Comment by IchordeDionysos** [comment]
@deepak786 I do appreciate your work you put in tough! Really nice if support comes some time 😍

**Comment by Zazo032** [comment]
Overall changes look good and reasonable to me. However, I think we should publish the platform interface as a separate package so it can be updated independently. What do you think @v3rm0n?

**Comment by gabdsg** [comment]
Any update on this?

**Comment by Zazo032** [comment]
> Any update on this?

To get this published to pub.dev, we will have to do 2 separate things, maybe @deepak786 can help by splitting this into 2 different PRs.

1. The first PR will deal with plugin federation, by moving the plugin folder and creating the platform interface.
2. After that platform interface gets published (cc/ @v3rm0n), another PR with the web implementation can be merged and publish the web-compatible plugin version to pub.dev

**Comment by creativecreatorormaybenot** [comment]
@Zazo032 @v3rm0n a thought from my side: it looks like the **contribution** here is essentially finished.

As you have highlighted, the missing steps are properly integrating it (creating a platform interface package etc.).  
That said, it seems like this is something the maintainer can do as well as it is trivial and depends mainly on publishing decisions.

Long story short: I would suggest you merge the contribution and then go through steps 1. and 2. yourself as you should be the ones deciding about it :)

**Comment by deepak786** [comment]
@Zazo032 @v3rm0n I think there is no need to create a separate package for the platform interface. 
Please check the package printing (https://github.com/DavBfr/dart_pdf/tree/master/printing). It also doesn't have a separate package for the platform interface.

You guys can proceed to merge.

**Comment by creativecreatorormaybenot** [comment]
@deepak786 It is not **required** but it is best practice because it allows you to properly work with multiple platforms.

**Comment by deepak786** [comment]
Created the separated pull request https://github.com/v3rm0n/intercom_flutter/pull/137
1. Moved the intercom_flutter code to intercom_flutter/intercom_flutter
2. Created intercom_flutter_platform_interface package.

The next steps would be 
1. first review and merge this pull request https://github.com/v3rm0n/intercom_flutter/pull/137
2. publish the package intercom_flutter_platform_interface on pub.dev
3. migrate intercom_flutter to use intercom_flutter_platform_interface

And then we can proceed with the web version. (I will create a separate package for that).

**Comment by deepak786** [comment]
Tasklist:

- [x] Create a separate package intercom_flutter_platform_interface. (https://github.com/v3rm0n/intercom_flutter/pull/137) 
- [x] Merge the pull request https://github.com/v3rm0n/intercom_flutter/pull/137
- [x] Push the package intercom_flutter_platform_interface to pub.dev
- [x] Create a separate package intercom_flutter_web (Web implementation) using intercom_flutter_platform_interface. (https://github.com/v3rm0n/intercom_flutter/pull/138)
- [x] Merge the pull request for intercom_flutter_web (https://github.com/v3rm0n/intercom_flutter/pull/138)
- [x] Migrate intercom_flutter to use intercom_flutter_platform_interface (https://github.com/v3rm0n/intercom_flutter/pull/139)
- [x] push the package intercom_flutter_web to pub.dev
- [x] add intercom_flutter_web in intercom_flutter (https://github.com/v3rm0n/intercom_flutter/pull/153)
- [x] Update intercom_flutter (5.0.0) probably on pub.dev

**Comment by deepak786** [comment]
@Zazo032 let me know when the intercom_flutter_web package is published.

**Comment by Zazo032** [comment]
@v3rm0n should push it to pub

**Comment by IchordeDionysos** [comment]
@v3rm0n @Zazo032 can you mark the intercom_flutter_platform_interface unlisted on pub.dev.

This is usually done to reduce clutter on pub.dev also as you should in almost all cases not directly depend on the platform interface :) 

image

**Comment by Zazo032** [comment]
> @v3rm0n @Zazo032 can you mark the intercom_flutter_platform_interface unlisted on pub.dev.
> 
> This is usually done to reduce clutter on pub.dev also as you should in almost all cases not directly depend on the platform interface :)
> 
> image

Done, didn't know that option existed. I guess web implementation shouldn't be unlisted since it can be used on its own (maybe a web-only Flutter app)

**Comment by IchordeDionysos** [comment]
Puh, I am super unsure here :D 

I've looked at how [`firebase_auth`](https://pub.dev/packages/firebase_auth) handles it and they have the [`firebase_auth_web`](https://pub.dev/packages/firebase_auth_web) listed.

But [`url_launcher`](https://pub.dev/packages/url_launcher) has any other [`_linux`](https://pub.dev/packages/url_launcher_linux), [`_macos`](https://pub.dev/packages/url_launcher_macos), [`_web`](https://pub.dev/packages/url_launcher_web) unlisted. (But [`_windows`](https://pub.dev/packages/url_launcher_windows) is listed 🙃)

And for [`connectivity`](https://pub.dev/packages/connectivity) is [`_for_web`](https://pub.dev/packages/connectivity_for_web) listed, but [`_macos`](https://pub.dev/packages/connectivity_macos) unlisted.
_By the way it triggers me that they called it_ `_for_web` 😂

In the design doc for federated plugins they say under ["Search results clutter"](https://docs.google.com/document/u/1/d/1LD7QjmzJZLCopUrFAAE98wOUQpjmguyGTN2wd_89Srs/mobilebasic#h.z1zte5tff6yr):

> In the monolithic-plugins world the above collapses into a single result: webview_flutter, which is most likely (unless there are platform specific features, see below) the only package the user cares about.

So if the web plugin would add additional functionality it would be listed 🤔

**Comment by Zazo032** [comment]
As I understand, we should unlist the web plugin once it's shipped (federated) into the intercom_flutter plugin

**Comment by deepak786** [comment]
is the intercom_flutter_web is published on pub.dev?

**Comment by Zazo032** [comment]
> is the intercom_flutter_web is published on pub.dev?

Not yet, waiting for @v3rm0n to publish it

**Comment by deepak786** [comment]
@Zazo032 @v3rm0n can't we create a group/organization on pub.dev in which we will be the members and anyone from us can upload/update the package?

**Comment by Zazo032** [comment]
I was invited by @v3rm0n but probably creating a team would be a better option to manage this plugin, since it's now federated

**Comment by Nico04** [comment]
Any idea when this can/will be merged & published ? Thanks for your work :)

**Comment by deepak786** [comment]
@Nico04 You can check the progress here https://github.com/v3rm0n/intercom_flutter/pull/111#issuecomment-826933240

Still waiting for the package intercom_flutter_web to be uploaded on pub.dev

**Comment by v3rm0n** [comment]
Sorry guys I haven't used Flutter for a few years now so let me know what exactly I need to run or what accesses should I grant so that things would not be blocked by me being unavailable.

**Comment by deepak786** [comment]
@v3rm0n Please grant access to Pub.dev so that we can upload or update the packages. Maybe you need to create an organization there.


But most important for now, please upload the package `intercom_flutter_web` on the pub.dev

**Comment by creativecreatorormaybenot** [comment]
@v3rm0n You should probably transfer the role of maintainer to someone else.

**Comment by deepak786** [comment]
@v3rm0n I created a publisher [intercom.deepakdroid.xyz](https://pub.dev/publishers/intercom.deepakdroid.xyz/packages). You can transfer the intercom flutter packages to this publisher. I will add you and @Zazo032 as a member of the publisher. So that anyone from us can update the package.

**Comment by gabdsg** [comment]
Any update on this? Thanks

**Comment by deepak786** [comment]
@v3rm0n Please share your email address so that I can add you as a member in the publisher.

**Comment by v3rm0n** [comment]
Transfer initiated

**Comment by deepak786** [comment]
`intercom_flutter: 5.0.0` is released with web support.
Don't hesitate to buy me a coffee https://www.buymeacoffee.com/deepakdroid :)

**Pull request closed without merging**